simulator/simulator.py

```diff
@@ -0,0 +1,386 @@
+        '''
+
+        cart_position, cart_velocity, *_ = self.client.getJointState(
+            bodyUniqueId=self.object_id,
+            jointIndex=self.slider_to_cart_index
+        )
+
+        pole_angle, pole_velocity, *_ = self.client.getJointState(
+            bodyUniqueId=self.object_id,
+            jointIndex=self.cart_to_pole_index
+        )
+
+        return State(
+            position=cart_position,
+            velocity=cart_velocity,
+            acceleration=self.target_acceleration,
+            pole_angle=pole_angle,
+            pole_angular_velocity=pole_velocity,
+            error_code=self.error
+        )
+
+    def get_target(self) -> float:
+        '''
+        Current acceleration targeet of system.
+        '''
+
+        return self.target_acceleration
+
+    def get_config(self) -> Config:
+        '''
+        Current config.
+        '''
+
+        assert self.config
+        return self.config
+
+    def reset(self, config: Config) -> None:
+        '''
+        Resets the environment to an initial state.
+        The pole is at rest position and cart is centered.
+        '''
+        log.info(f'reset config {config}')
+
+        assert self.physical_params
+        if config.clamp_velocity or config.clamp_position:
+            raise NotImplementedError
+
+        self.config = config
+        self.target_acceleration = 0
+        self.velocity = 0
+        self.delta_velocity = 0
+        self.step_count = 0
+
+        self.client.resetJointState(
+            bodyUniqueId=self.object_id,
+            jointIndex=self.cart_to_pole_index,
+            targetValue=0,
+            targetVelocity=0,
+        )
+
+        self.client.resetJointState(
+            bodyUniqueId=self.object_id,
+            jointIndex=self.slider_to_cart_index,
+            targetValue=0,
+            targetVelocity=0,
+        )
+
+        self.client.setJointMotorControl2(
+            bodyUniqueId=self.object_id,
+            jointIndex=self.slider_to_cart_index,
+            controlMode=pb.VELOCITY_CONTROL,
+            targetVelocity=0,
+            force=self.physical_params.motor_force
+        )
+
+        self.error = Error.NO_ERROR
+
+    def __set_motor_velocity(self, velocity) -> None:
+        '''
+        Set joint velocity
+        '''
+
+        log.debug(f'set motor velocity {velocity}')
+
+        self.client.setJointMotorControl2(
+            bodyUniqueId=self.object_id,
+            jointIndex=self.slider_to_cart_index,
+            controlMode=pb.VELOCITY_CONTROL,
+            targetVelocity=velocity,
+        )
+
+    def set_target(self, target: float) -> None:
+        '''
+        Controll acceleration of cart.
+        It is assumed that the torque is sufficient to execute the command.            
+
+        Args:
+            * target – desired acceleration of cart m/s
+        '''
+
+        if bool(self.error):
+            log.warning(f'set target, error={self.error}')
+            return
```

Comment:
The concept is just raise flag without any exceptions, because we can reset cart pole and continue  ¯\_(ツ)_/¯

**P.S.** Hope, device has the same behavior.